### PR TITLE
Make max length of _id field configurable.

### PR DESCRIPTION
**Pull request by howardhuanghua**
This PR going to make max length of _id field configurable. This issue has been discussed in https://github.com/elastic/elasticsearch/pull/16036.

As a Cloud Elasticsearch Service provider, some of our cloud customers also complain about this hardcode limitation. In their case, they dump data from many types of database to ES directly, including their own-developed database, and use ES to provide search capability. The original "rowkey" or "id" limit length in their database may be more than 512 and use it as ES _id field directly, as they need doc idempotence. In this case they could not import few of the data entries due to this limitation. They know the huge _id field performance issue and HTTP GET issue, they could accept these issues and just want customer data could be saved and searchable.

We could also see people in community complain this hardcode limitation recently: https://discuss.elastic.co/t/id-length/182245/7

So I think we need to make this hardcode limitation configurable and just tell customer the disadvantages if you want to increase the limitation.

In this PR, I introduced a new dynamic cluster setting, keep 512 as default value:
`PUT /_cluster/settings
{
  "transient": {
    "indices.max.doc_id.length": 1024
  }
}`
Customer could dynamically update the limitation if they encounter this index issue.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/indices/IndicesService.java [hunk 1]
```diff
@@ -190,6 +191,10 @@
         Setting.Property.NodeScope
     );
 
+    public static int MAX_DOC_ID_LENGTH = 512;
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/indices/IndicesService.java:194]: We try hard to avoid static variables, since this hampers test-ability and leads to unstable tests. For instance in this PR, all tests that run in the same JVM after `IndexRequestTests` will run with a limit of 513.

I believe this requires doing the validation outside the request, as also explained in [#16036](https://github.com/elastic/elasticsearch/pull/16036#issuecomment-186897826).

**Review by henningandersen — CHANGES_REQUESTED** [review]
@howardhuanghua thanks for looking into this. I think more changes are necessary to implement this fully:

1. We should implement this without using static variable to hold the dynamic state.
2. I believe Lucene still has a limit of 32766 bytes, we should verify that and put that in as an upper limit to avoid allowing too large values.
3. I would prefer to make this an unmodifiable setting on an index. This allows us to optimize for the case where this is not set in the future. Also, it requires an explicit opt-in rather than just up'ing the limit when it is hit. It feels like something index specific rather than something you want to apply to all your indices to avoid the issues of not being able to use GET requests to retrieve back the docs.
4. Documentation will need to be amended both where it describes the existing limit and to describe the new setting.

Also, can you elaborate on why the workaround of using a hash and storing the actual id in a separate field will not work in your case(s)?

**Comment by howardhuanghua** [comment]
Hi @henningandersen, thanks for your suggestion. I our customer's case, they think using a hash and storing the actual id in a separate field could not completely avoid hash collision, and also need extra data transformations in their data sync flow.

I have moved the max doc id limitation config to index level. Here are the related changes:
1. Introduce `index.max_doc_id_length` dynamic config in index setting. I make it dynamic since if user found some documents have been limited by max doc id length, they could dynamically update the index setting without re-indexing or creating a new index. This would be more convenient for users.
2. I moved the limitation validation outside of `IndexRequest` and put it in `TransportShardBulkAction` .  I have checked that both single document index request and bulk operations would go through this action. And I also only validate doc id length on primary shard.
3. The max doc id length setting is limited by Lucene’s max term length which is 32766 bytes.
4. Documents have been updated in both index and bulk API.

Please help to review again. Thank you.

**Comment by howardhuanghua** [comment]
Hi @henningandersen , would please help to review again? Thanks a lot.

**Comment by howardhuanghua** [comment]
Hi @henningandersen , would please help to review again? Thanks a lot.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java [hunk 2]
```diff
@@ -269,8 +272,24 @@ static boolean executeBulkItemRequest(BulkPrimaryExecutionContext context, Updat
                 request.ifSeqNo(), request.ifPrimaryTerm());
         } else {
             final IndexRequest request = context.getRequestToExecute();
+            String id = request.id();
+            IndexSettings indexSettings = context.getPrimary().indexSettings();
+            if (indexSettings != null) {
+                int maxDocIdLen = context.getPrimary().indexSettings().getMaxDocIdLength();
+                if (id != null &&
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:279]: I think we expect id to be non-null at this time, i.e., we should rather assert that than make it part of the condition.

File: server/src/main/java/org/elasticsearch/index/IndexSettings.java [hunk 3]
```diff
@@ -72,6 +73,8 @@
             (value) -> Translog.Durability.valueOf(value.toUpperCase(Locale.ROOT)), Property.Dynamic, Property.IndexScope);
     public static final Setting<Boolean> INDEX_WARMER_ENABLED_SETTING =
         Setting.boolSetting("index.warmer.enabled", true, Property.Dynamic, Property.IndexScope);
+    public static final Setting<Integer> INDEX_MAX_DOC_ID_LENGTH_SETTING =
+        Setting.intSetting("index.max_doc_id_length", 512, 1, Property.Dynamic, Property.IndexScope);
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/index/IndexSettings.java:77]: Using `doc_id` is unfortunate since the lucene "doc id" is something else, suggest to just stick to `id`:

```suggestion
        Setting.intSetting("index.max_id_length", 512, 1, Property.Dynamic, Property.IndexScope);
```

(also rename various fields and variables).

File: server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java [hunk 4]
```diff
@@ -269,8 +272,24 @@ static boolean executeBulkItemRequest(BulkPrimaryExecutionContext context, Updat
...
+                    id.getBytes(StandardCharsets.UTF_8).length > maxDocIdLen) {
+                    ActionRequestValidationException validationException = new ActionRequestValidationException();
+                    validationException.addValidationError("id is too long, must be no longer than "
+                        + maxDocIdLen + " bytes but was: " +
+                        id.getBytes(StandardCharsets.UTF_8).length);
+                    result = new Engine.IndexResult(validationException, request.version());
+                    context.markOperationAsExecuted(result);
+                    context.markAsCompleted(context.getExecutionResult());
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java:287]: This moves the error handling from request to item, meaning previously the entire bulk request would fail, whereas now, only the individual item will fail. I am not really opposed to this, but will need to double check whether we need backwards compatibility semantics on this change.

Likewise, this will also fail for update and delete I think? I would prefer to leave those out of this for now, since update/delete of a too long id will now error out instead of just being treated as "not found".

File: server/src/main/java/org/elasticsearch/common/settings/ClusterSettings.java [hunk 5]
```diff
@@ -203,6 +203,7 @@ public void apply(Settings value, Settings current, Settings previous) {
             IndicesQueryCache.INDICES_QUERIES_CACHE_ALL_SEGMENTS_SETTING,
             IndicesService.INDICES_ID_FIELD_DATA_ENABLED_SETTING,
             IndicesService.WRITE_DANGLING_INDICES_INFO_SETTING,
+            IndicesService.INDICES_CACHE_CLEAN_INTERVAL_SETTING,
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/common/settings/ClusterSettings.java:206]: This looks unrelated to the PR? I would prefer to revert that.

File: server/src/main/java/org/elasticsearch/index/IndexSettings.java (hunk 3 above)
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/index/IndexSettings.java:77]: The default value of 1 is a bit worrying. First of all, it gives very few docs possible. But also, it is smaller than our auto-id size. That would make for strange messages if using auto-generated ids.

I think I would prefer to just leave the minimum at 512. Or at least at the max-size of an auto-generated id.

Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/index/IndexSettings.java:77]: Rather than making this setting `Dynamic`, I think we should make it `Final` at least for now. As mentioned earlier, this could enable us to optimize for the default id length case in the future. Is there significant use cases you have, where dynamically updating this is necessary?

File: docs/reference/docs/index_.asciidoc [hunk 6]
```diff
@@ -99,7 +99,9 @@ The name of the index the document was added to.
 The document type. {es} indices now support a single document type, `_doc`.
 
 `_id`::
-The unique identifier for the added document.
+The unique identifier for the added document. By default, maximum length
```
Comment (henningandersen) [inline docs/reference/docs/index_.asciidoc:102]: I think mentioning the max limit on the index-modules page is enough, it feels a bit low-level to include here, since a lot of users go here early in their Elasticsearch endeavors.

**Review by henningandersen — CHANGES_REQUESTED** [review]
Hi @howardhuanghua sorry for the delay on getting back to you. I left a number of comments. I think we should also add an integration test, either an `ESIntegTestCase` or a YAML REST test. There is a couple of outstanding items I need to circle back on as well.

**Comment by henningandersen** [comment]
A few concerns have been raised on this:

1. The known issue with GET API and the `id` length.
2. Potential performance impact (and maybe increased OOM risk) with the `LiveVersionMap`
3. Similarly performance issues when using `update_by_query` and `delete_by_query`

I will discuss this further internally to try to reach a conclusion on whether we want to move ahead with this and/or if it has to be more restricted in use somehow.

It does highlight the need to be explicit about the tradeoffs in documentation (which you are welcome to work on adding into this PR).

**Comment by howardhuanghua** [comment]
Thank you @henningandersen , I will update it based on your comment, if you get some discussion details please help to share with me, thanks.

**Comment by pugnascotia** [comment]
@howardhuanghua we've discussed this change some more within the Elasticsearch team, and we don't feel that making the max length of the `_id` field is the right approach. Instead, we think it would be better to add support for generating hashes as part of an ingest pipeline. We've raised https://github.com/elastic/elasticsearch/issues/59633 to cover this work.

As such, I'm going to close this PR for now. I would like to thank you for the time you spent on this work.

**Comment by howardhuanghua** [comment]
@pugnascotia thanks. So we could create a ingest pipeline and using `sha1()` or `sha256()` to transfer `_id` field to a hex strings right? So how could we get doc by original id value?

**Comment by stu-elastic** [comment]
@howardhuanghua 
Store the original id in a different field.  There's an [example](https://github.com/elastic/elasticsearch/pull/59671#issuecomment-659467724) in the PR.  In that case `bar` is the original id.

**Comment by howardhuanghua** [comment]
Thanks @stu-elastic @pugnascotia @henningandersen . Use `sha256()` to transfer long `_id` field is a workaround. But we still could not entirely avoid hash collision caused duplicated id value right? It has the possible even it's very rare. But in database data-sync scenario, it would be fatal.

**Comment by stu-elastic** [comment]
If you find a collision for sha256 please give me an acknowledgement on the paper you publish announcing the result ;-)
